packages/xdl/src/project/publishAsync.ts

```diff
@@ -0,0 +1,459 @@
+      ExponentTools.getManifestAsync(response.url, {
+        'Exponent-SDK-Version': exp.sdkVersion,
+        'Exponent-Platform': 'android',
+        'Expo-Release-Channel': options.releaseChannel,
+        Accept: 'application/expo+json,application/json',
+      }),
+      ExponentTools.getManifestAsync(response.url, {
+        'Exponent-SDK-Version': exp.sdkVersion,
+        'Exponent-Platform': 'ios',
+        'Expo-Release-Channel': options.releaseChannel,
+        Accept: 'application/expo+json,application/json',
+      }),
+    ]);
+
+    const hookOptions = {
+      url: response.url,
+      exp,
+      iosBundle,
+      iosSourceMap,
+      iosManifest,
+      androidBundle,
+      androidSourceMap,
+      androidManifest,
+      projectRoot,
+      log: (msg: any) => {
+        logger.global.info({ quiet: true }, msg);
+      },
+    };
+
+    for (const hook of validPostPublishHooks) {
+      logger.global.info(`Running postPublish hook: ${hook.file}`);
+      try {
+        runHook(hook, hookOptions);
+      } catch (e) {
+        logger.global.warn(`Warning: postPublish hook '${hook.file}' failed: ${e.stack}`);
+      }
+    }
+  }
+
+  const fullManifestUrl = response.url.replace('exp://', 'https://');
+  await EmbeddedAssets.configureAsync({
+    projectRoot,
+    pkg,
+    exp,
+    releaseChannel: options.releaseChannel ?? 'default',
+    iosManifestUrl: fullManifestUrl,
+    iosManifest,
+    iosBundle,
+    iosSourceMap,
+    androidManifestUrl: fullManifestUrl,
+    androidManifest,
+    androidBundle,
+    androidSourceMap,
+    target,
+  });
+
+  // TODO: move to postPublish hook
+  // This method throws early when a robot account is used for a kernel build
+  if (exp.isKernel && user.kind !== 'robot') {
+    await _handleKernelPublishedAsync({
+      user,
+      exp,
+      projectRoot,
+      url: response.url,
+    });
+  }
+
+  return {
+    ...response,
+    url:
+      options.releaseChannel && options.releaseChannel !== 'default'
+        ? `${response.url}?release-channel=${options.releaseChannel}`
+        : response.url,
+  };
+}
+
+async function _uploadArtifactsAsync({
+  exp,
+  iosBundle,
+  androidBundle,
+  options,
+  pkg,
+}: {
+  exp: ExpoConfig;
+  iosBundle: string;
+  androidBundle: string;
+  options: PublishOptions;
+  pkg: PackageJSONConfig;
+}) {
+  logger.global.info('');
+  logger.global.info('Uploading JavaScript bundles');
+  const formData = new FormData();
+
+  formData.append('expJson', JSON.stringify(exp));
+  formData.append('packageJson', JSON.stringify(pkg));
+  formData.append('iosBundle', iosBundle, 'iosBundle');
+  formData.append('androidBundle', androidBundle, 'androidBundle');
+  formData.append('options', JSON.stringify(options));
+
+  const user = await UserManager.ensureLoggedInAsync();
+  const api = ApiV2.clientForUser(user);
+
+  return await api.uploadFormDataAsync('publish/new', formData);
+}
+
+async function _handleKernelPublishedAsync({
```

Comment:
Looking forward to getting this removed 🤞 

https://github.com/expo/expo/issues/9086